jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -2365,7 +2372,31 @@ public T newInstance() throws IllegalAccessException, InstantiationException {
 		checkNonSunProxyMemberAccess(security, callerClassLoader, Member.PUBLIC);
 	}
 
+/*[IF JAVA_SPEC_VERSION >= 12]*/
+	Class<?> callerClazz = getStackClass(1);
+	if (callerClazz.classLoader == ClassLoader.bootstrapClassLoader) {
+		/* newInstanceImpl() is required for all bootstrap classes to avoid an infinite loop
+		 * when calling copyConstructor.invoke() in cacheConstructor() at the bootstrap stage
+		 * as the constructors of bootstrap classes are not yet cached for use at that time.
+		 */
+		return (T)J9VMInternals.newInstanceImpl(this);
+	} else {
+		try {
+			Constructor<?> ctr = getDeclaredConstructor();
+			if (reflectionFactory == null) {
+				reflectionFactory = AccessController.doPrivileged(new ReflectionFactory.GetReflectionFactoryAction());
```

Comment:
See the `writeFieldValue()` helper. It should be used to set the field so this is thread safe.

Reply:
Updated with `writeFieldValue()` against the suggestion above.